File: tests_C++/test1.cpp
/**
 * Nullity test of Divergence operator
 */

#include "mole.h"
#include <iostream>

void run_nullity_test(int k, Real tol) {
  int m = 2 * k + 1;
  Real dx = 1;

  Divergence D(k, m, dx);
  vec field(m + 1, fill::ones);

  vec sol = D * field;

  if (norm(sol) > tol) {
    cout << "\033[1;31mTest FAILED!\033[0m\n";
    exit(1);
  }
}

int main() {
  Real tol = 1e-10;

  for (int k : {2, 4, 6})
    run_nullity_test(k, tol);

  cout << "\033[1;32mTest PASSED!\033[0m\n";

  return 0;
}

File: tests_MATLAB/test1.m
% Nullity test of Divergence operator

addpath('../mole_MATLAB')

ks = [2, 4, 6, 8];  % Different orders of accuracy
tol = 1e-10;

for k = ks
    m = 2 * k + 1;
    dx = 1 / m;

    D = div(k, m, dx);
    
    field = ones(m + 1, 1);
    
    sol = D * field;
    
    if (norm(sol) > tol)
        fprintf("Test FAILED!\n");
    end
end

fprintf("Test PASSED!\n");
